docs/snapshotting/snapshot-support.md

```diff
@@ -0,0 +1,300 @@
+in multiple files that are composing the full microVM snapshot:
+- the guest memory file,
+- the microVM state file,
+- zero or more disk files (depending on how many the guest had; these are
+  **managed by the users**, which means they need to externally back up their
+  block devices backing files).
+
+The design allows sharing of memory pages and read only disks between multiple
+microVMs. When loading a snapshot, instead of loading at resume time the full
+contents from file to memory, Firecracker creates a
+[MAP_PRIVATE mapping](http://man7.org/linux/man-pages/man2/mmap.2.html) of the
+memory file, resulting in runtime on-demand loading of memory pages. Any subsequent
+memory writes go to a copy-on-write anonymous memory mapping.
+This has the advantage of very fast snapshot loading times, but comes with the cost
+of having to keep the guest memory file around for the entire lifetime of the
+resumed microVM.
+
+## Firecracker Snapshotting characteristics
+
+- Fresh Firecracker microVMs are booted using `anonymous` memory, while microVMs
+  resumed from snapshot load memory on-demand from the snapshot and copy-on-write
+  to anonymous memory.
+- Resuming from a snapshot is optimized for speed, while taking a snapshot involves
+  some extra CPU cycles for synchronously writing dirty memory pages to the memory
+  snapshot file. Taking a snapshot of a fresh microVM, on which dirty pages tracking
+  is not enabled, results in the full contents of guest memory being written to the
+  snapshot.
+- The _memory file_ and _microVM state file_ are generated by Firecracker on snapshot
+  creation, and the disk contents are flushed to their backing files.  
+- The API calls exposing the snapshotting functionality have clear **Prerequisites**
+  that describe the requirements on when/how they should be used.
+ 
+## Snapshot versioning
+
+Firecracker snapshotting implementation offers support for microVM versioning
+(`cross-version snapshots`) in the following contexts:
+- saving snapshots at older versions (being able to create a snapshot with any version
+in the `[N, N + o]` interval, while being in Firecracker version `N+o`),
+- loading snapshots from older versions (being able to load a snapshot created by any
+Firecracker version in the `[N, N + o]` interval, in a Firecracker version `N+o`).
+
+The design supports an unlimited number of versions, the value of `o` (maximum number
+of older versions that we can restore from / save a snapshot to, from the current
+version) will be defined later.
+
+## Snapshot API
+
+Firecracker exposes the following APIs for manipulating snapshots: `Pause`, `Resume`
+and `CreateSnapshot` can be called only after booting the microVM, while `LoadSnapshot`
+is allowed only before boot.
+
+### Pausing the microVM
+
+To create a snapshot, first you have to pause the running microVM and its vCPUs with
+the following API command:
+
+```bash
+curl --unix-socket /tmp/firecracker.socket -i \
```

Comment:
While using `curl`, the outer quotes must be singular and the inner ones must be double.